From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Cui <cuizhihui030925@outlook.com>
Date: Fri, 15 Jan 2021 19:58:40 +0800
Subject: [PATCH] Swallow API


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 10cab1b0939d54bcd2b351c2317baf5ce5e97684..c60081368e15f6d51ba7f5fdd95ab3040a464c07 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1520,6 +1520,13 @@ public interface Server extends PluginMessageRecipient {
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Yatopia end
+        // Swallow start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSwallowConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Swallow end
 
         /**
          * Sends the component to the player
